cmd/bcpd/app/msg.go

```diff
@@ -0,0 +1,55 @@
+package app
+
+import (
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/errors"
+	"github.com/iov-one/weave/x/batch"
+)
+
+var _ batch.Msg = (*BatchMsg)(nil)
+
+func (*BatchMsg) Path() string {
+	return batch.PathExecuteBatchMsg
+}
+
+func (msg *BatchMsg) MsgList() ([]weave.Msg, error) {
+	messages := make([]weave.Msg, len(msg.Messages))
+	// make sure to cover all messages defined in protobuf
+	for i, m := range msg.Messages {
+		res, err := func() (weave.Msg, error) {
+			switch t := m.GetSum().(type) {
+			case *BatchMsg_Union_SendMsg:
+				return t.SendMsg, nil
+			case *BatchMsg_Union_NewTokenMsg:
+				return t.NewTokenMsg, nil
+			case *BatchMsg_Union_SetNameMsg:
+				return t.SetNameMsg, nil
+			case *BatchMsg_Union_CreateEscrowMsg:
+				return t.CreateEscrowMsg, nil
+			case *BatchMsg_Union_ReleaseEscrowMsg:
+				return t.ReleaseEscrowMsg, nil
+			case *BatchMsg_Union_ReturnEscrowMsg:
+				return t.ReturnEscrowMsg, nil
+			case *BatchMsg_Union_UpdateEscrowMsg:
+				return t.UpdateEscrowMsg, nil
+			case *BatchMsg_Union_CreateContractMsg:
+				return t.CreateContractMsg, nil
+			case *BatchMsg_Union_UpdateContractMsg:
+				return t.UpdateContractMsg, nil
+			case *BatchMsg_Union_SetValidatorsMsg:
+				return t.SetValidatorsMsg, nil
+			default:
+				return nil, errors.ErrUnknownTxType(t)
+			}
+		}()
+		if err != nil {
+			return messages, err
```

Comment:
If you want to return messages together with error, it's better to create the result slice as `make([]weave.Msg, 0, len(msg.Messages))` - this way you won't have nil messages in it.